File: zero-downtime-migration/src/routes/v1Users.ts
// Import Router from Express to create modular route handlers
import { Router } from "express";

// Import the User model to interact with the users table in the DB
import User from "../models/User";

// Create a new Express router instance
const router = Router();

// ------------------------
// 1. POST /users
// Create a new user in the database
// ------------------------
router.post("/users", async (req, res) => {
  // Extract fields from request body
  const { full_name, email } = req.body;

  // Create a new user in the database
  // Maps full_name from API request to fullName in the database
  const user = await User.create({
    fullName: full_name,
    email
  });

  // Respond with the newly created user as JSON
  res.json(user);
});

// ------------------------
// 2. GET /users/:id
// Fetch a user by primary key
// ------------------------
router.get("/users/:id", async (req, res) => {
  // Find user by primary key (id) from URL parameter
  const user = await User.findByPk(req.params.id);

  // Respond with a simplified JSON object
  // Use optional chaining in case user is not found
  res.json({
    id: user?.id,           // User ID or undefined if not found
    full_name: user?.fullName // User fullName or undefined
  });
});

// Export the router to be mounted in main app
export default router;

File: zero-downtime-migration/python/routes/v1_users.py
# routes/v1_users.py
# Defines version 1 of the Users API.
# This version represents the OLD contract that existing clients still rely on.
# It continues to use the legacy `full_name` field to avoid breaking consumers
# during a zero-downtime migration.

from fastapi import APIRouter
from database import SessionLocal
from models import User

# Create a router for v1 endpoints
router = APIRouter()

@router.post("/users")
def create_user_v1(full_name: str, email: str):
    # Create a new database session for this request
    db = SessionLocal()

    # OLD API behavior:
    # Existing clients only send `full_name`, not first_name / last_name.
    # We continue writing to the legacy column so old clients keep working.
    user = User(full_name=full_name, email=email)

    # Persist the new user record
    db.add(user)
    db.commit()

    # Refresh the instance to load generated fields (e.g., id)
    db.refresh(user)

    # Return response in the OLD API format
    return {
        "id": user.id,
        "full_name": user.full_name
    }

@router.get("/users/{user_id}")
def get_user_v1(user_id: int):
    # Create a database session
    db = SessionLocal()

    # Fetch user by primary key
    user = db.query(User).filter(User.id == user_id).first()

    # Return only legacy fields expected by v1 clients
    return {
        "id": user.id,
        "full_name": user.full_name
    }
